### ISSUE TITLE
How to add to esphome?

### ISSUE THREAD

**Issue by abishur** [PaulWieland/ratgdo#21]
Is there anyway to add the install configuration files to esphome so I can manage this device along side my other ESPhome devices?

**Comment by jcam** [comment]
The project would need to be modified quite a bit to pull out just the door control stuffs as sensors and switches, split from the wifi/mqtt/etc stuff that ESPhome handles itself. If this continues to not get traction, maybe I'll poke at it once my shields arrive.

**Comment by kantlivelong** [comment]
Not requiring ESPHome is great but I agree that an ESPHome compatible library/component would be really cool.

Once the 2.x source is pushed out I'll likely take a look at what adjustments can be made to make the code compatible with both.

**Comment by bdraco** [comment]
I tried to port it to esphome but I couldn't get it to read the UART correctly.  I might have the pins wrong though

https://github.com/bdraco/esphome-ratgdo

**Comment by bdraco** [comment]
```
---
esphome:
  name: ratgdo
  platform: ESP8266
  board: esp01_1m

api:
  id: api_server

external_components:
  - source:
      type: git
      url: https://github.com/bdraco/esphome-ratgdo
    refresh: 1s

uart:
  tx_pin: 2
  rx_pin: 4
  baud_rate: 9600

ratgdo:

# WiFi connection, correct these
# with values for your WiFi.
wifi:
  ssid: <fill>
  password: <fill>

logger:
#  level: VERBOSE


# Enable over-the-air updates.
ota:

# Sync time with Home Assistant.
time:
  - platform: homeassistant
    id: homeassistant_time


```

**Comment by kantlivelong** [comment]
> I've got the esphome lib reading state now.

Nice! Is this written for the V1 or V2 board?

**Comment by bdraco** [comment]
```
[08:50:07][V][mdns:112]:   - _http, _tcp, 80
[08:50:07][C][ota:093]: Over-The-Air Updates:
[08:50:07][C][ota:094]:   Address: overlookratgdo.local:8266
[08:50:07][C][api:138]: API Server:
[08:50:07][C][api:139]:   Address: overlookratgdo.local:6053
[08:50:07][C][api:143]:   Using noise encryption: NO
[08:50:11][D][ratgdo:429]: Motion state 1
[08:50:26][D][ratgdo:413]: Door state 4

```

**Comment by bdraco** [comment]
<img width="1021" alt="Screenshot 2023-06-07 at 12 28 06 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/fd537528-8c73-4ce7-8582-9d6da5aba4db">


Getting close to usable

**Comment by kantlivelong** [comment]
@bdraco Do you think it would be possible to merge into this repo in a way that allows the dedicated firmware and esphome to use the same codebase? Obviously we are still waiting on 2.x source and @PaulWieland would have to be open to the idea but I think that would be a great way to give users choice while using the upstream source.

**Comment by bdraco** [comment]
Working esphome config for shield v2 -- note, you must currently factory reset the GDO before installing since there is no way to set or change the rolling code and it will start at 0

The light is still a bit wonky but will work that out

```
---
esphome:
  name: overlookratgdo
  platform: ESP8266
  board: esp01_1m

api:
  id: api_server

web_server:

external_components:
  - source:
      type: git
      url: https://github.com/bdraco/esphome-ratgdo
    refresh: 1s

ratgdo:
  id: overlook_ratgdo

binary_sensor:
  - platform: ratgdo
    type: motion
    id: ratgdo_motion
    ratgdo_id: overlook_ratgdo
    name: "Overlook Garage Motion"
    device_class: motion
  - platform: ratgdo
    type: obstruction
    id: ratgdo_obstruction
    ratgdo_id: overlook_ratgdo
    name: "Overlook Garage Obstruction"
    device_class: problem

cover:
  - platform: ratgdo
    id: ratgdo_garage
    device_class: garage
    name: "Overlook Garage"
    ratgdo_id: overlook_ratgdo


light:
  - platform: ratgdo
    id: ratgdo_light
    name: "Overlook Garage Light"
    ratgdo_id: overlook_ratgdo

uart:
  tx_pin: 
      number: 2
      inverted: true
  rx_pin:
      number: 4
      inverted: true
  baud_rate: 9600

wifi:
  ssid: YOURSSID
  password: YOURPASSWORD

logger:
  level: VERBOSE

ota:

# Sync time with Home Assistant.
time:
  - platform: homeassistant
    id: homeassistant_time


```

**Comment by bdraco** [comment]
everything is working well now.  I'm working on adding support for a sync button and a input number to change the counter

**Comment by bdraco** [comment]
<img width="999" alt="Screenshot 2023-06-07 at 5 29 49 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/cadacfc4-fe44-4e52-abd6-d4c308c5e394">

**Comment by jcam** [comment]
Looking great!!

**Comment by bdraco** [comment]
<img width="1032" alt="Screenshot 2023-06-07 at 6 31 55 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/86e4b68f-5c6b-442a-bf32-351084b66e5b">
<img width="1013" alt="Screenshot 2023-06-07 at 6 31 48 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/0cc00e2e-cdfa-4c19-a18d-0de4ba20bc2b">


You can now set the counter so easy to transition to the esphome firmware

**Comment by bdraco** [comment]
light state is still a bit flakey for some reason

sync and query buttons implemented


<img width="879" alt="Screenshot 2023-06-07 at 7 14 36 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/5d3933c6-87ee-4ecd-9bf3-5a81e42ce875">
<img width="1253" alt="Screenshot 2023-06-07 at 7 14 23 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/eb14043c-7752-4caf-9c7e-7bd65d797f31">

**Comment by bdraco** [comment]
I think I've ironed out all the issues. I'm going to move it into production for testing

**Comment by bdraco** [comment]
<img width="775" alt="Screenshot 2023-06-07 at 9 52 24 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/107a1c00-1cb9-41d0-bc16-dff214c80f43">

Installed in production and looking good

**Comment by bdraco** [comment]
> @bdraco Do you think it would be possible to merge into this repo in a way that allows the dedicated firmware and esphome to use the same codebase? Obviously we are still waiting on 2.x source and @PaulWieland would have to be open to the idea but I think that would be a great way to give users choice while using the upstream source.

I rewrote a huge chunk of the code to get it to work with esphome.

Everything would need to be broken out into a new library to share code which might be more trouble than it’s worth considering how different esphome works.  I replaced the serial library with the one built into esphome.  The only library code it shares is sec plus

It’s easy enough to setup a page to flash them if people are interested.  

The code is all at the repo above so feel free to copy use / etc.

**Comment by bdraco** [comment]
If anyone does decide to play around with this, don’t forget to save the rolling code counter before flashing

**Comment by bdraco** [comment]
<img width="315" alt="Screenshot 2023-06-08 at 1 44 48 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/16914a6f-cfa2-4561-8de3-8f8cd41db5da">

https://esphome-ratgdo.github.io/esphome-ratgdo/

I put up a flasher above if anyone wants to play with it.

You can always reflash back to official firmware as long as you save the rolling counter value

**Comment by kantlivelong** [comment]
I plan on trying it out once I mod my V1 boards to V2.

**Comment by bdraco** [comment]
I made it adoptable in the esphome dashboard so once you have it flashed you can manage updates from there

<img width="494" alt="Screenshot 2023-06-08 at 2 05 24 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/2e74f639-4e60-4ed5-8629-fd6ea04f7bb9">

**Comment by bdraco** [comment]
<img width="520" alt="Screenshot 2023-06-09 at 7 41 51 PM" src="https://github.com/PaulWieland/ratgdo/assets/663432/51f10ea4-1911-4301-a421-69efe7666563">


Added remote lock out, openings, button sensor, dry connect sensors

**Comment by Automayted** [comment]
Excellent work, @bdraco!

**Comment by robgridley** [comment]
@bdraco 🙇‍♂️

**Comment by bdraco** [comment]
I deployed to 14 garage doors now and all working well.

I fixed a few bugs and changed the config around a bit so if you already flashed, please reflash

I also added esp32 support for the wemos d1 esp32 in case anyone wants a chip with more ram to do noise encryption

Stock board
https://raw.githubusercontent.com/ESPHome-RATGDO/esphome-ratgdo/main/static/v2board_esp8266_d1_mini_lite.yaml

Esp32
https://raw.githubusercontent.com/ESPHome-RATGDO/esphome-ratgdo/main/static/v2board_esp32_d1_mini.yaml

**Comment by bdraco** [comment]
I pulled the esp32 support because there are some problems to be worked out still and I didn’t want anyone to waste their time with it.  I’ll put it back up when they are solved.

**Comment by bdraco** [comment]
> @bdraco amazing work. question: how do you "save the value of the rolling code counter and set it again" when going from stock firmware to your esphome build?

Check the console log of the stock firmware.  It prints the code whenever you execute an action.  The esphome version has a number entity that allows you to change the stored value

**Comment by jfroy** [comment]
Thanks a lot @bdraco !

**Comment by jfroy** [comment]
Just flashed to the ESPHome firmware successfully. Everything seems to be working.

**Comment by kantlivelong** [comment]
Installed today - working great!

Thanks so much!

**Comment by bdraco** [comment]
There were some new features added this week if you flashed the initial version.

The docs now show how to migrate https://github.com/ESPHome-RATGDO/esphome-ratgdo#moving-from-stock-ratgdo

**Comment by efaden** [comment]
This looks awesome!  The move to position will be really cool.  Actually built a hacky version of that using a aqara vibration sensor to measure the door angle..

**Comment by Buzzzsaw** [comment]
I just installed mine today using your ESPHome port, and it seems to be working great so far!
Big props for this :)

**Comment by BreeOge** [comment]
This needs to be stickied somehow, thank you bdraco!
